src/microphone/browser_audio_processor.ts

```diff
@@ -0,0 +1,117 @@
+import { ErrDeviceNotSupported, ErrNoAudioConsent, ErrNotInitialized } from './types'
+
+export type AudioHandler = (audioBuffer: Float32Array) => void
+
+export interface AudioProcessor {
+  readonly sampleRate: number
+  initialize(): Promise<void>
+  close(): void
+  mute(): void
+  unmute(): void
+}
+
+const audioProcessEvent = 'audioprocess'
+
+export class BrowserAudioProcessor implements AudioProcessor {
+  private initialized: boolean = false
+  readonly sampleRate: number
+
+  // Audio callback invoked when audio is returned by the script processor.
+  private readonly isSafari: boolean
+  private readonly onAudio: AudioHandler
+  private readonly audioContext: AudioContext
+
+  // The media stream and audio track are initialized during `initialize()` call.
+  private audioTrack?: MediaStreamTrack
+  private mediaStream?: MediaStream
+
+  // Audio processing functionality is initialized lazily during first `unmute()` call.
+  // The reason for that is that it's required for user to first interact with the page,
+  // before it can capture or play audio and video, for privacy and user experience reasons.
+  private audioProcessor?: ScriptProcessorNode
+
+  constructor(onAudio: AudioHandler) {
+    if (window.AudioContext !== undefined) {
+      this.audioContext = new window.AudioContext()
+      this.isSafari = false
+    } else if (window.webkitAudioContext !== undefined) {
+      // eslint-disable-next-line new-cap
+      this.audioContext = new window.webkitAudioContext()
+      this.isSafari = true
+    } else {
+      throw ErrDeviceNotSupported
+    }
+
+    this.sampleRate = this.audioContext.sampleRate
+    this.onAudio = onAudio
+  }
+
+  async initialize(): Promise<void> {
+    if (window.navigator?.mediaDevices === undefined) {
+      throw ErrDeviceNotSupported
+    }
+
+    if (this.isSafari) {
```

Comment:
Could you add a comment on why this is done differently on Safari?

Reply:
Will do!

Reply:
Added the comments to both branches!